Made the UI for expiry

manojmalik20 opened the pull request:
Made expiry menu component with all the expiry items
Made the dateTimeInput component with the react-day-picker along with customized navbar
Removed the memoization from the getCustomStatus selector
Solved the propType issue in the menuItemToggleModalRedux

<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

#### Related Pull Requests
<!--
List all PRs related to resolving a ticket. For instance, if you submitted a PR to `mattermost/mattermost-server`, please include it here.
-->
- Has server changes (please link here)
- Has redux changes (please link here)
- Has mobile changes (please link here)

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.
-->

manojmalik20 committed:
Made the UI for expiry
Made expiry menu component with all the expiry items
Made the dateTimeInput component with the react-day-picker along with customized navbar
Removed the memoization from the getCustomStatus selector
Solved the propType issue in the menuItemToggleModalRedux

chetanyakan commented on components/custom_status/custom_status_modal.tsx:
```diff
@@ -56,9 +58,11 @@ const CustomStatusModal: React.FC<Props> = (props: Props) => {
     const [showEmojiPicker, setShowEmojiPicker] = useState<boolean>(false);
     const [text, setText] = useState<string>(currentCustomStatus.text || '');
     const [emoji, setEmoji] = useState<string>(currentCustomStatus.emoji);
+    const [expiry, setExpiry] = useState<string>('four-hours');
```
Please define and use constants.

chetanyakan commented on components/custom_status/date_time_input.tsx:
```diff
@@ -0,0 +1,151 @@
+                    }
+                }}
+            >
+                <svg
+                    width='10'
+                    height='16'
+                    viewBox='0 0 10 16'
+                    fill='none'
+                    xmlns='http://www.w3.org/2000/svg'
+                >
+                    <path
+                        d='M9.878 1.91605L8.528 0.566049L1.094 8.00005L8.528 15.434L9.878 14.066L3.794 8.00005L9.878 1.91605ZM9.5 1.84405L8.15 0.494048L0.5 8.00005L8.15 15.506L9.5 14.156L3.2 8.00005L9.5 1.84405Z'
+                        fill='#3D3C40'
+                        fillOpacity='0.32'
+                    />
+                </svg>
```
Check if this can be replaced with font awesome icons

chetanyakan commented on components/custom_status/date_time_input.tsx:
```diff
@@ -0,0 +1,151 @@
+
+    const handleDayChange = (day: Date, modifiers: DayModifiers) => {
+        if (modifiers.today) {
+            setDay('Today');
```
Use constants as much as possible.

chetanyakan commented on components/custom_status/date_time_input.tsx:
```diff
@@ -0,0 +1,151 @@
+                        aria-hidden='true'
+                    />
+                </span>
+                <input value='12:30 PM'/>
```
For the custom date/time field, the next time value should be the time next 30th minute or hour, based on the current time.

chetanyakan commented on components/custom_status/expiry_menu.tsx:
```diff
@@ -0,0 +1,104 @@
+                                <Menu.ItemAction
+                                    key={index}
+                                    onClick={(event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => handleExpiryChange(event, item)}
+                                    ariaLabel={localizeMessage(`expiry_dropdown.${expiryMenuItems[item].localizationId}`, expiryMenuItems[item].text).toLowerCase()}
+                                    text={localizeMessage(`expiry_dropdown.${expiryMenuItems[item].localizationId}`, expiryMenuItems[item].text)}
```
1. Using template string for localization ids causes the i18n script to remove these strings. Instead, we should define the entire ID in the `expiryMenuItems` constant.

2. Using the same ariaLabel and text seems redundant. Check if that is recommended practice for a11y.

chetanyakan commented on selectors/views/custom_status.ts:
```diff
@@ -11,14 +11,18 @@ import {UserCustomStatus} from 'mattermost-redux/types/users';
 
 import {GlobalState} from 'types/store';
 
+// const timer = 1616155500110;
```
Remove commented code

chetanyakan commented on selectors/views/custom_status.ts:
```diff
@@ -11,14 +11,18 @@ import {UserCustomStatus} from 'mattermost-redux/types/users';
+        const userProps = user?.props || {};
+        const rv = userProps.customStatus ? JSON.parse(userProps.customStatus) : undefined;
+
+        // if (Date.now() < timer) {
+        return rv;
+
+        // }
```
Remove commented code

chetanyakan commented on selectors/views/custom_status.ts:
```diff
@@ -11,14 +11,18 @@ import {UserCustomStatus} from 'mattermost-redux/types/users';
+    return (state: GlobalState, userID?: string) => {
+        const user = userID ? getUser(state, userID) : getCurrentUser(state);
+        const userProps = user?.props || {};
+        const rv = userProps.customStatus ? JSON.parse(userProps.customStatus) : undefined;
+
+        // if (Date.now() < timer) {
+        return rv;
```
```suggestion
        return userProps.customStatus ? JSON.parse(userProps.customStatus) : undefined;
```

chetanyakan commented on selectors/views/custom_status.ts:
```diff
@@ -11,14 +11,18 @@ import {UserCustomStatus} from 'mattermost-redux/types/users';
 import {GlobalState} from 'types/store';
 
+// const timer = 1616155500110;
 export function makeGetCustomStatus(): (state: GlobalState, userID?: string) => UserCustomStatus {
```
Use a normal selector `getCustomStatus` instead of using `makeGetCustomStatus` function that returns a selector

manojmalik20 committed:
Review fixes and added timepicker menu
Changed the custom status selector to a simple function and changed its use everywhere
Made constants for the expiry menu items and used them
Changed the expiry menu items to use localized text
Removed the daypicker.scss file and did the changes in _daypicker.scss file
Added all the localization ids in en.json

manojmalik20 committed:
Completed the timepicker menu
Made a util for getting current date and time for timezone
Made the function to round current time to nearest 30 min time
Made the function to fill the menu with 30 minute intervals
Unified the separate state variables for day and time to only time
Made the dateTimeInput component to be controlled by the custom status modal
Fixed the styles for the timepicker and daypicker
Fixed lint errors and type check errors

manojmalik20 committed:
Fixed some issues in daypicker
Fixed the styling in daypicker
Added logic for disabled days in daypicker
Fixed the issue of prevMonth button not working in daypicker

manojmalik20 committed:
Complete expiry support in custom status modal
Changed the custom status type to include duration and expires_at and made enum duration
Removed the duration constants from constants file and used enum type everywhere
Added localization in the expiry menu values
Modified custom status modal to send expiry with the custom status
Modified custom status modal to handle initial custom expiry time
Modified the disable set status variable to handle the confirm button disability
Modified the custom status selector to return custom status only when it's not expired

manojmalik20 committed:
Review fixes
Code refactoring in custom status modal and dateTime input and expiry menu

manojmalik20 committed:
Changed name of enum Duration to CustomStatusDuration

manojmalik20 committed:
Merge pull request #49 from brightscout-alpha/MI-1221_1
